### ISSUE TITLE
Breaks window selection from overview by keyboard

### ISSUE THREAD

**Issue by bbo2adwuff** [p91paul/middleclickclose#18]
When middleclickclose (branch 3.38) is enabled I cannot select window from Overview with the keyboard anymore.

#### Steps to reproduce

1. Hit Super.
2. Navigate to window with arrow keys.
3. Hit Enter

When I disable middleclickclose then it works again.

(I also reported it here https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/3259)

**Comment by p91paul** [comment]
Fixed in latest master branch, uploading it to e.g.o.
Thanks for the bug report.

**Comment by bbo2adwuff** [comment]
Wow thanks for fixing it so quick :rocket: 
The bug report on gitlab.gnome.org is closed already since 2 days ;)

Thanks!

**Comment by p91paul** [comment]
The fix for this brought worse issues (#20). I need to fix my fix, meanwhile I unreleased v18 on extensions.gnome.org.

**Comment by gitbobbed** [comment]
Hey Paul,

Has there been any progress in fixing this bug? Were you able to fix the fix?

Also, thanks for the great work. It's been hard to adapt to GNOME without your extension, but I had to because being able to select a window with the keyboard while in overview is super important for me.

Best,

Daniel

**Comment by bbo2adwuff** [comment]
Now in GNOME 40 at least graphically it seems that I can navigate with the arrow keys to certain windows in the overview.

Try following steps:
1. Hit Super.
2. Navigate to window with arrow keys.
3. Hit Enter

1 and 2 work, only the `Enter` in steps is not working.

I don't know if that helps or if that was the case already before that 2 works!?

**Comment by bbo2adwuff** [comment]
I really don't have any clue about Gnome extensions, but wouldn't the following logic work?

You use `if button == middleclick` here:
https://github.com/p91paul/middleclickclose/blob/7c1653bf00da0bc28296ce921cc79ccf4d91e6d4/middleclickclose%40paolo.tranquilli.gmail.com/extension.js#L75

What about an extra line `if button == enter`?

I mean it's so simple that you probably thought about it already!?

**Comment by p91paul** [comment]
@bbo2adwuff IIRC that handles mouse events only, so I guess it's a no :disappointed:

**Comment by vjxyz** [comment]
I also have this issue.. can't select a window by pressing Enter from overview if I have the extension installed 😢

**Comment by neun24** [comment]
From reading the code with no knowledge about Gnome extensions I think you're overriding WindowClone's activate and thus overriding it completely. Doesn't matter if you're using mouse or keyboard from now on, it just won't get activated anymore.

```
		// override WindowClone's _activate
		WindowPreview.prototype._activate = () => {};
```

**Comment by pesader** [comment]
Hi everyone! I did some digging while trying to fix this and found out a few things.

## The cause

This extension overrides the [`_activate()`](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/750ade34dae296511c246dc6bb07442120141841/js/ui/windowPreview.js#L554) function making it empty.

```javascript
WindowPreview.prototype._activate = () => {};
```

As a result, when it is called by the function that handles the Enter key press nothing happens. Below is the [code from GNOME Shell](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/750ade34dae296511c246dc6bb07442120141841/js/ui/windowPreview.js#L597) that handles that:

```javascript
  vfunc_key_press_event(keyEvent) {
      let symbol = keyEvent.keyval;
      let isEnter = symbol == Clutter.KEY_Return || symbol == Clutter.KEY_KP_Enter;
      if (isEnter) {
          this._activate(); // this is overriden as empty, so nothing happens
          return true;
      }

      return super.vfunc_key_press_event(keyEvent);
  }
```

## Attempt 1

To fix this, I tried overriding the `vfunc_key_press_event(keyEvent)` itself to use the `_oldActivate()` function that this extension stores upon being enabled. This way, it would call the original `_activate`, not the one overridden as empty.

```javascript
// save the original key press handling function
this._oldVfuncKeyPressEvent = WindowPreview.prototype.vfunc_key_press_event;


// override the key press handling function
WindowPreview.prototype.vfunc_key_press_event = function(keyEvent) {
    let symbol = keyEvent.keyval;
    let isEnter = symbol == Clutter.KEY_Return || symbol == Clutter.KEY_KP_Enter;
    if (isEnter) {
	init._oldActivate.apply(this);
	return true;
    }

    // if it wasn't the Enter key, use the original function
    return this._oldVfuncKeyPressEvent(keyEvent);
}
```

But that change made the shell crash whenever the activities view was shown with one window or more. Anyways, you can test it yourself on [this branch](https://github.com/pesader/middleclickclose/tree/override-key-press-handler).

## Attempt 2

I then realized there would be no need to override the `_activate()` function if we could simply disconnect the `'clicked'` signal handler. So here's what I did:

```javascript
// override _addWindowClone to add my event handler
Workspace.Workspace.prototype._addWindowClone = function(metaWindow) {
    let clone = init._oldAddWindowClone.apply(this, [metaWindow]);

    // remove default 'clicked' signal handler
    let id = GObject.signal_handler_find(
        clone.get_actions()[0],
        {signalId: 'clicked'}
    )
    clone.get_actions()[0].disconnect(id);

    // add custom 'clicked' signal handler
    clone.get_actions()[0].connect('clicked', onClicked.bind(clone));

    return clone;
}
```

But that also lead to the same crashing on overview 🙃
I have a branch with these changes [here](https://github.com/pesader/middleclickclose/tree/disconnect-clicked-handler).

## Conclusion

I think I'm on to something but I don't know what I'm doing, so I just hope maybe this is something @p91paul will able to build from 😅

**Comment by p91paul** [comment]
attempt 2 looks promising. Can you retrieve which error happens exactly from journalctl logs?

**Comment by pesader** [comment]
> Can you retrieve which error happens exactly from journalctl logs?

Of course! Which command should I run to do that?

**Comment by p91paul** [comment]
The ideal way is to run `journalctl -f`, leave it running, reproduce the issue, then kill journalctl with ctrl+c and copy the output text.
If the crash makes you unable to use a terminal afterwards, but you manage to restore the system to a working state without rebooting, `journalctl -b` gives the logs from boot, but make sure to only get the last few minutes because it will be huge.
If you are forced to reboot, journalctl also has options to get logs from previous boot.

**Comment by p91paul** [comment]
you reference GObject in your code for attempt 2, but you probably didn't import it. At the top of the file you see a series of imports, you need to check where that GObject is defined and import it (if you copied the code from somewhere else, they'll probably have such an import statement

**Comment by pesader** [comment]
> You reference GObject in your code for attempt 2, but you probably didn't import it.

Yep, that was the problem. It's working now, so I'll make a PR.

> If you copied the code from somewhere else, they'll probably have such an import statement

I wrote it myself! 🤓 
(That's probably why there was such a silly mistake in it)

Thanks for the help!

**Comment by p91paul** [comment]
Nice job, thanks!

**Comment by p91paul** [comment]
Fixed by @pesader via #41, I submitted it to extensions.gnome.org and I expect it to be approved shortly. Thanks!
